### ISSUE TITLE
Code generation does not take `@include`/`@skip` into account for named fragment optionality

### ISSUE THREAD

**Issue by designatednerd** [apollographql/apollo-ios#1678]
Generally when fields are accessed with an `@include` or `@skip` directive that takes a parameter, we represent those fields as optional. This falls apart when those directives are applied to a named fragment, such as [in our `HeroDetailsFragmentConditionalInclusionQuery`](https://github.com/apollographql/apollo-ios/blob/fa07e3957232173eb2604b811493f1acf431e91b/Sources/StarWarsAPI/API.swift#L2298).

The fragment is still generated as non-nullable, and because any force-unwrapping takes place under the hood rather than at the level of property access, attempting to do an unwrap such as `if let name = hero.fragments.heroDetails.name` will still crash at runtime. 

The real fix is to make sure we're generating these fragments as optional on the `Fragments` property. This may or may not be plausible to fix in existing codegen, we'll have to do some investigation. 

A workaround is to check the `hero.fragments.heroDetails.resultMap` to see how many items are in it - in testing only the `__typename` was included and other properties were omitted. 

Big thanks to @Ariel-Steinlauf-okcupid for bringing this to our attention.

**Comment by russellporter** [comment]
Glad to see this is on the radar, we hit what is I think the same root issue when using an `@include` directive on a union type field.

Here's a contrived example where we fetch different data on the same node from two separate fragments:
```graphql
query ReproBug($imageIdsEnabled: Boolean! = false) {
  posts {
    ...PostImageURLs
    ...PostImageIDs
  }
}

fragment PostImageURLs on Launch {
  media {
    ... on Image {
      url
    }
  }
}

fragment PostImageIDs on Launch {
  media @include(if: @imageIdsEnabled)) { # type: `[Image | Video]!`
    ... on Image {
      id
    }
  }
}
```

Trying to access the image id would then crash, when `@imageIdsEnabled` is false.

`posts.fragments.postImageIDs.media.first?.asImage?.id`  <- crashes when force unwrapping `id`. The "Image" object is there because we fetched the URL for it in the other fragment.

**Comment by designatednerd** [comment]
oy. ok good to know this is also an issue on unions (cc @AnthonyMDev)

**Comment by calvincestari** [comment]
Hi 👋🏻 - this issue should be resolved in 1.0. 1.0 is currently in alpha releases with a full 1.0 release in H2 2022. Please take a look at the [latest alpha](https://github.com/apollographql/apollo-ios/releases) and let us know if this is still an issue for you.

**Comment by russellporter** [comment]
Awesome, thank you! I am not using Apollo on my current project but this is great to hear that this issue is solved.
